### PR TITLE
Add and remove dependent services correctly in hierarchical structures

### PR DESCRIPTION
**Pull request by d-m-u**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1515741

Changes how we add and remove dependent services in hierarchical structures. Previously dependent services weren't being retired because we weren't setting the hierarchy correctly.

### CODE REVIEW

**Comment by d-m-u** [comment]
@miq-bot add_label blocker, bug
@miq-bot assign @gmcculloug

File: lib/miq_automation_engine/service_models/miq_ae_service_service.rb [hunk 1]
```diff
@@ -76,9 +76,9 @@ def parent_service=(service)
         if service
           raise ArgumentError, "service must be a MiqAeServiceService" unless service.kind_of?(
             MiqAeMethodService::MiqAeServiceService)
-          @object.service = Service.find(service.id)
+          @object.add_to_service(service)
         else
-          @object.service = nil
+          @object.remove_from_service(service.parent)
```
Comment (gmcculloug) [inline lib/miq_automation_engine/service_models/miq_ae_service_service.rb]: You want to pass `@object.parent` to the method.

**Comment by d-m-u** [comment]
Changes how we add and remove dependent services in hierarchical structure for https://bugzilla.redhat.com/show_bug.cgi?id=1515741. Previously dependent services weren't being retired because we weren't setting the hierarchy correctly.

File: spec/service_models/miq_ae_service_service_spec.rb [hunk 2]
```diff
@@ -81,6 +83,7 @@ def invoke_ae
 
         @parent.reload
         expect(@parent.direct_service_children).to eq([])
+        expect(@parent.service_resources).to eq([])
       end
 
       it "validates the parent service" do
```
Comment (tinaafitz) [inline spec/service_models/miq_ae_service_service_spec.rb]: @d-m-u The changes look good. Could we add a test for when the child Service being added to the Service already has a parent Service? We would expect it to raise an error.

File: spec/service_models/miq_ae_service_service_spec.rb [hunk 3]
```diff
@@ -72,6 +72,20 @@ def invoke_ae
...
+      it "raises error if child service is already owned by different parent" do
+        @parent1 = FactoryGirl.create(:service, :name => "parent_service1")
+        method = "$evm.root['service'].parent_service = $evm.vmdb('service').find(#{@parent1.id})"
+        @ae_method.update_attributes(:data => method)
+        invoke_ae
+
+        method = "$evm.root['service'].parent_service = $evm.vmdb('service').find(#{@parent.id})"
+        @ae_method.update_attributes(:data => method)
+
+        expect { invoke_ae }.to raise_error(MiqAeException::AbortInstantiation)
```
Comment (gmcculloug) [inline spec/service_models/miq_ae_service_service_spec.rb]: This test could be written to use the service model directly without having to call `invoke_ae` which would be quicker and simpler while still validating the same use case: 

Something like:
```ruby
      it "raises error if child service is already owned by different parent" do
        parent1 = FactoryGirl.create(:service)
        service.add_to_service(parent1)
        parent_service = MiqAeMethodService::MiqAeServiceService.find(@parent.id)

        expect { service_service.parent_service = parent_service }.to raise_error(MiqException::Error)
      end
```

**Comment by simaishi** [comment]
I assume `gaprindashvili/yes`?

**Comment by d-m-u** [comment]
@miq-bot add_label gaprindashvili/yes

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit 66426386f97805d0996348e7e714a88a02a3b4af
Author: Greg McCullough <gmccullo@redhat.com>
Date:   Thu Dec 14 07:54:57 2017 -0500

    Merge pull request #132 from d-m-u/bz_1515741
    
    Add and remove dependent services correctly in hierarchical structures
    (cherry picked from commit 6c5f4fff48f169888da419bc49247652152a4b2b)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1526458
```

**Pull request merged**